### PR TITLE
Feature/integer index pow

### PR DESCRIPTION
**Pull request by vbkaisetsu**
Implementation of #26

### CODE REVIEW

File: python-primitiv/primitiv/_graph.pyx [hunk 1]
```diff
@@ -148,7 +148,9 @@ cdef class _Node:
     def __pow__(left, right, mod):
         if mod is not None:
             return NotImplemented
-        if isinstance(right, (int, float)):
+        if isinstance(right, int) and right >= 0:
```
Comment (odashi) [inline python-primitiv/primitiv/_graph.pyx]: The value assertion about `right` should be fixed:
* It should accept negative integers.
* It should reject overflow-ed values because Python3 does not distinguish 32-bits integers and longer integers.

File: python-primitiv/primitiv/_operator.pyx [hunk 2]
```diff
@@ -98,7 +98,9 @@ class _operators:
 
     @staticmethod
     def pow(x, k):
-        if isinstance(x, _Node) and isinstance(k, (int, float)):
+        if isinstance(x, _Node) and isinstance(k, int) and k >= 0:
```
Comment (odashi) [inline python-primitiv/primitiv/_operator.pyx]: ditto

File: python-primitiv/primitiv/_operator.pyx [hunk 3]
```diff
@@ -377,7 +379,9 @@ class _tensor_operators:
 
     @staticmethod
     def pow(x, k):
-        if isinstance(x, _Tensor) and isinstance(k, (int, float)):
+        if isinstance(x, _Tensor) and isinstance(k, int) and k >= 0:
```
Comment (odashi) [inline python-primitiv/primitiv/_operator.pyx]: ditto

File: python-primitiv/primitiv/_tensor.pyx [hunk 4]
```diff
@@ -160,7 +160,9 @@ cdef class _Tensor:
     def __pow__(left, right, mod):
         if mod is not None:
             return NotImplemented
-        if isinstance(right, (int, float)):
+        if isinstance(right, int) and right >= 0:
```
Comment (odashi) [inline python-primitiv/primitiv/_tensor.pyx]: ditto

File: python-primitiv/primitiv/_operator.pyx [hunk 5]
```diff
@@ -98,7 +98,9 @@ class _operators:
 
     @staticmethod
     def pow(x, k):
-        if isinstance(x, _Node) and isinstance(k, (int, float)):
+        if isinstance(x, _Node) and isinstance(k, int) and -0x80000000 <= k <= 0x7fffffff:
```
Comment (odashi) [inline python-primitiv/primitiv/_operator.pyx]: ditto

File: python-primitiv/primitiv/_operator.pyx [hunk 6]
```diff
@@ -377,7 +379,9 @@ class _tensor_operators:
 
     @staticmethod
     def pow(x, k):
-        if isinstance(x, _Tensor) and isinstance(k, (int, float)):
+        if isinstance(x, _Tensor) and isinstance(k, int) and -0x80000000 <= k <= 0x7fffffff:
```
Comment (odashi) [inline python-primitiv/primitiv/_operator.pyx]: ditto

File: python-primitiv/primitiv/_tensor.pyx [hunk 7]
```diff
@@ -160,7 +160,9 @@ cdef class _Tensor:
     def __pow__(left, right, mod):
         if mod is not None:
             return NotImplemented
-        if isinstance(right, (int, float)):
+        if isinstance(right, int) and -0x80000000 <= right <= 0x7fffffff:
```
Comment (odashi) [inline python-primitiv/primitiv/_tensor.pyx]: ditto

File: python-primitiv/tests/node_operators.py [hunk 8]
```diff
@@ -79,4 +79,15 @@ def test_node_pow(self):
...
         self.assertTrue(np.isclose((2 ** x).to_ndarrays()[0], np.array([[2, 4], [8, 16]])).all())
+        self.assertTrue(np.isclose((x ** -2).to_ndarrays()[0], np.array([[1, 1/4], [1/9, 1/16]])).all())
+        input_arr = np.array([1, -1, 3, -3, 5, -5])
+        x = F.input(input_arr)
+        self.assertTrue(((x ** 6).to_ndarrays()[0] == np.array([1, 1, 729, 729, 15625, 15625])).all())
+        self.assertTrue(((x ** 9).to_ndarrays()[0] == np.array([1, -1, 19683, -19683, 1953125, -1953125])).all())
+        input_arr = np.array([1, -1])
+        x = F.input(input_arr)
+        self.assertTrue(((x ** 0x7fffffff).to_ndarrays()[0] == np.array([1, -1])).all())
+        self.assertTrue(((x ** -0x80000000).to_ndarrays()[0] == np.array([1, 1])).all())
```
Comment (odashi) [inline python-primitiv/tests/node_operators.py]: Fix corner cases.

File: python-primitiv/tests/tensor_operators.py [hunk 9]
```diff
@@ -76,6 +76,17 @@ def test_tensor_pow(self):
...
         self.assertTrue(np.isclose((2 ** x).to_ndarrays()[0], np.array([[2, 4], [8, 16]])).all())
+        self.assertTrue(np.isclose((x ** -2).to_ndarrays()[0], np.array([[1, 1/4], [1/9, 1/16]])).all())
+        input_arr = np.array([1, -1, 3, -3, 5, -5])
+        x = tF.input(input_arr)
+        self.assertTrue(((x ** 6).to_ndarrays()[0] == np.array([1, 1, 729, 729, 15625, 15625])).all())
+        self.assertTrue(((x ** 9).to_ndarrays()[0] == np.array([1, -1, 19683, -19683, 1953125, -1953125])).all())
+        input_arr = np.array([1, -1])
+        x = tF.input(input_arr)
+        self.assertTrue(((x ** 0x7fffffff).to_ndarrays()[0] == np.array([1, -1])).all())
+        self.assertTrue(((x ** -0x80000000).to_ndarrays()[0] == np.array([1, 1])).all())
```
Comment (odashi) [inline python-primitiv/tests/tensor_operators.py]: ditto

File: test/tensor_ops_test.cc [hunk 10]
```diff
@@ -1142,6 +1142,56 @@ TEST_F(TensorOpsTest, CheckLog) {
...
+    0.000001, 0.125, 1, 8, 64, 512,
+    -0.000001, -0.125, -1, -8, -64, -512,
+  };
+  for (Device *dev : devices) {
+    const Tensor x = dev->new_tensor_by_vector(Shape({2, 3}, 2), x_data_1);
+    const Tensor y = ipow(x, 3);
+    EXPECT_EQ(Shape({2, 3}, 2), y.shape());
+    EXPECT_TRUE(vector_match(y_data_1, y.to_vector()));
+  }
+  const vector<float> x_data_2 {
```
Comment (odashi) [inline test/tensor_ops_test.cc]: No difference with `x_data_1`.

File: test/tensor_ops_test.cc [hunk 11]
```diff
@@ -1142,6 +1142,56 @@ TEST_F(TensorOpsTest, CheckLog) {
...
+  };
+  for (Device *dev : devices) {
+    const Tensor x = dev->new_tensor_by_vector(Shape({2, 3}, 2), x_data);
+    const Tensor y = pow(x, 2.5);
+    EXPECT_EQ(Shape({2, 3}, 2), y.shape());
+    EXPECT_TRUE(vector_match(y_data, y.to_vector()));
+  }
+}
+
+TEST_F(TensorOpsTest, CheckPowInt) {
```
Comment (odashi) [inline test/tensor_ops_test.cc]: Might better to separate two test cases.

**Comment by vbkaisetsu** [comment]
Boundary check is only fixed in C++, not in Python.
